orttraining/orttraining/training_api/include/onnxruntime_training_cxx_api.h

```diff
@@ -0,0 +1,151 @@
+#pragma once
+#include "onnxruntime_training_c_api.h"
+#include <optional>
+
+namespace Ort {
+
+#define ORT_DECLARE_TRAINING_RELEASE(NAME) \
+  void OrtRelease(Ort##NAME* ptr);
+
+// These release methods must be forward declared before including onnxruntime_cxx_api.h
+// otherwise class Base won't be aware of them
+ORT_DECLARE_TRAINING_RELEASE(CheckpointState);
+ORT_DECLARE_TRAINING_RELEASE(TrainingSession);
+
+}  // namespace Ort
+
+#include "onnxruntime_cxx_api.h"
+
+namespace Ort {
+
+inline const OrtTrainingApi& GetTrainingApi() { return *GetApi().GetTrainingApi(ORT_API_VERSION); }
+
+#define ORT_DEFINE_TRAINING_RELEASE(NAME) \
+  inline void OrtRelease(Ort##NAME* ptr) { GetTrainingApi().Release##NAME(ptr); }
+
+ORT_DEFINE_TRAINING_RELEASE(CheckpointState);
+ORT_DEFINE_TRAINING_RELEASE(TrainingSession);
+
+class CheckpointState : Base<OrtCheckpointState> {
+ private:
+  friend class TrainingSession;
+  friend CheckpointState LoadCheckpoint(const std::basic_string<ORTCHAR_T>& path_to_checkpoint);
+
+  CheckpointState(OrtCheckpointState* checkpoint_state) { p_ = checkpoint_state; }
+
+ public:
+  // Construct the checkpoint state by loading the checkpoint by calling LoadCheckpoint
+  CheckpointState() = delete;
+};
+
+/** \brief Manage the training loop using this class
+ *
+ * Wraps OrtTrainingSession
+ *
+ */
+class TrainingSession : public Base<OrtTrainingSession> {
+ private:
+  friend void SaveCheckpoint(const TrainingSession& session, const std::basic_string<ORTCHAR_T>& path_to_checkpoint,
+                             bool include_optimizer_states);
+
+  size_t training_model_output_count_, eval_model_output_count_;
+
+ public:
+  TrainingSession(const SessionOptions& session_options, CheckpointState& checkpoint_state,
+                  const std::basic_string<ORTCHAR_T>& train_model_path,
+                  const std::optional<std::basic_string<ORTCHAR_T>>& eval_model_path =
+                      std::optional<std::basic_string<ORTCHAR_T>>(),
```

Comment:
I am not sure if it is fair to expect users to use c++17+. Since we are anyways using default arguments, I can remove `std::optional` and make those basic_strings default to being empty. If empty, the user did not provide it is a reasonable assumption to make i guess.
Hvaing said that, these apis are subject to change. We are not committing to these (atleast for now). So, can change in the near future.

Let me know if you have any other ideas.

Reply:
After syncing with @askhade, I think we are ok with using C++17.